lib/class-wp-theme-json-gutenberg.php

```diff
@@ -1073,6 +1073,7 @@ public function merge( $incoming ) {
 		// For leaf values that are arrays it will use the numeric indexes for replacement.
 		// In those cases, we want to replace the existing with the incoming value, if it exists.
 		$to_replace   = array();
+		$to_replace[] = array( 'layout', 'units' );
```

Comment:
@youknowriad would like confirmation about how `layout.units` should work. My understanding is that it should work this way:

- if a theme doesn't provide any `layout.units` defaults are listed
- themes can override these values via `theme.json`
- if a theme provides an empty array in `layout.units`, the control only shows `px`
- if a theme provides a limited set of units, such as `[ 'px', 'em' ]`, those are the only units shown in the control

Is this behavior correct?

If so, we need to add `layout.units` to the list of things we merge ourselves.

Reply:
@aristath would know more here but it does sound right.

Reply:
Though that makes me wonder, is `layout.units` a think? Why "units" are in "layout" I feel like that's an error?

Reply:
isn't it `spacing.units` ?

Reply:
Both exist and are used in different places: 

- `layout.units`: layout hook, letter-spacing component, border width hook, column width.
- `spacing.units`: padding hook, margin hook, cover height, unit control.

Reply:
just saw that there's a number of `layout.units` useSettings calls in the codebase, it feels we should replace these with `spacing.units`. I guess this was just a typo. cc @aristath

Reply:
ok, substituted `layout.units` per `spacing.units` at https://github.com/WordPress/gutenberg/pull/33280/commits/2d194400f9c79923a25fbf73df93da68cd286965 would welcome confirmation about this change

Reply:
Should `layout.units` & `spacing.units` be the same thing?
"Spacing" refers mostly to padding & margins, while "Layout" would refer to things like the width of a group block, the width of a column etc. Using `vh` in spacing would make sense, but it wouldn't make sense as the width of a column.
They are different scenarios, and "spacing" doesn't quite cover things that are not related to spacing.

Maybe we could just add a fallback to `layout.units` and make it use `spacing.units` if undefined? :thinking: 
This was part of PR https://github.com/WordPress/gutenberg/pull/31822#issue-644597488

Reply:
I think the difference between "width" and "height" is not the same thing as "layout" and "spacing", both can be considered layout. I think potentially some distinction between more specific settings (margins, paddings, widths, heights...) might be needed but the current difference doesn't make sense to me.

Reply:
The other side effect here is that we shouldn't add any config to `layout` that is not about the "layout config" itself (content width, wide width, type of layout), because that config has a meaning holistically and it's not a discrete list of settings.

Reply:
I think replacing `layout.units` with `spacing.units` would be OK... Both implementations are sub-optimal, but until we figure out a proper way to differentiate these things we should probably keep things as they were :+1: 
If and when the time comes to differentiate them, a separate ticket should be opened and we can investigate the implementation separately. For now, let's replace things with `spacing.units` and keep it consistent :heart:

Reply:
`spacing.units` can be overridden by block, so I wonder if that's enough for now. 

Another thing I've noticed is that the `UnitControl` uses `spacing.units` for `availableUnits`: it looks like these are the "allowed units" the control can use unless a consumer filters them via `units`. This means that what you pass as `units` can only be a subset of `spacing.units`: you can't have in `layout.units` things that are not part of `spacing.units` as well.

I'm also in favor of starting small and add things if we need them later. Remove things is hard/impossible.

Reply:
I haven't seen this in use by any theme and doesn't have documentation either, so I presume `layout.units` wasn't widely known/used in practice.

Reply:
For example, if you set `spacing.units` to an empty array, any content in `layout.units` would've been ignored. I think I'm going to commit this as it is. It's code, so we can always add more, should it be necessary.

Reply:
> The other side effect here is that we shouldn't add any config to layout that is not about the "layout config" itself (content width, wide width, type of layout), because that config has a meaning holistically and it's not a discrete list of settings.

Prepared this https://github.com/WordPress/gutenberg/pull/33303 to specify what settings can be part of layout, so we don't allow random things there.